### Importing the Libraries that are necessary to Train an Test the Data

In [1]:
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.naive_bayes import GaussianNB
from sklearn.metrics import classification_report


### Loading the necessary glass data to apply the algorithm

In [2]:
data = pd.read_csv('glass.csv')  # Replace 'glass.csv' with your dataset file path


In [3]:
data.head()

RI     Na    Mg    Al     Si     K    Ca   Ba   Fe  Type
0  1.52101  13.64  4.49  1.10  71.78  0.06  8.75  0.0  0.0     1
1  1.51761  13.89  3.60  1.36  72.73  0.48  7.83  0.0  0.0     1
2  1.51618  13.53  3.55  1.54  72.99  0.39  7.78  0.0  0.0     1
3  1.51766  13.21  3.69  1.29  72.61  0.57  8.22  0.0  0.0     1
4  1.51742  13.27  3.62  1.24  73.08  0.55  8.07  0.0  0.0     1

In [4]:
data.describe()

RI          Na          Mg          Al          Si           K  \
count  214.000000  214.000000  214.000000  214.000000  214.000000  214.000000   
mean     1.518365   13.407850    2.684533    1.444907   72.650935    0.497056   
std      0.003037    0.816604    1.442408    0.499270    0.774546    0.652192   
min      1.511150   10.730000    0.000000    0.290000   69.810000    0.000000   
25%      1.516522   12.907500    2.115000    1.190000   72.280000    0.122500   
50%      1.517680   13.300000    3.480000    1.360000   72.790000    0.555000   
75%      1.519157   13.825000    3.600000    1.630000   73.087500    0.610000   
max      1.533930   17.380000    4.490000    3.500000   75.410000    6.210000   

               Ca          Ba          Fe        Type  
count  214.000000  214.000000  214.000000  214.000000  
mean     8.956963    0.175047    0.057009    2.780374  
std      1.423153    0.497219    0.097439    2.103739  
min      5.430000    0.000000    0.000000    1.000000  
25%      8.240000    0.000000    0.000000    1.000000  
50%      8.600000    0.000000    0.000000    2.000000  
75%      9.172500    0.000000    0.100000    3.000000  
max     16.190000    3.150000    0.510000    7.000000

In [5]:
print(data.isnull().sum())  # Check for null values
data.fillna(data.mean(), inplace=True)  # Replacing all the null values with the mean data for the particular column

RI      0
Na      0
Mg      0
Al      0
Si      0
K       0
Ca      0
Ba      0
Fe      0
Type    0
dtype: int64


### Seperating the data in the form of Train and Test where 80% training data and 20% testing data

In [100]:
X = data.drop('Type', axis=1)
y = data['Type']
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=11)


In [6]:
from sklearn.preprocessing import StandardScaler #Normalization
scaler = StandardScaler()
X_train = scaler.fit_transform(X_train)
X_test = scaler.transform(X_test)


NameError: name 'X_train' is not defined

### Using the Algorithm as per the given question i.e., Naive Bayes Classifier **Train

In [101]:
nb_classifier = GaussianNB()
nb_classifier.fit(X_train, y_train)


GaussianNB()

### Predicting the Outcome ** Test

In [102]:
y_pred = nb_classifier.predict(X_test)


In [103]:
print(X_test.shape[0],(y_test!=y_pred).sum())

43 16


### Finding the accuracy of the model to find how correct the prediction is if accuracy is low follow various Normalization techniques

In [104]:
accuracy = round(nb_classifier.score(X_test, y_test)*100,2)
print(f"Accuracy of the Naïve Bayes model: {accuracy:.2f}")


Accuracy of the Naïve Bayes model: 62.79


### Finally Generating the report on how well the performace of the Algorithm is by looking for the Classification report

In [105]:
report = classification_report(y_test, y_pred, zero_division=1)
print("Classification Report:\n", report)

Classification Report:
               precision    recall  f1-score   support

           1       0.79      0.55      0.65        20
           2       0.50      0.56      0.53         9
           3       0.36      0.80      0.50         5
           5       1.00      0.00      0.00         2
           6       0.67      1.00      0.80         2
           7       1.00      1.00      1.00         5

    accuracy                           0.63        43
   macro avg       0.72      0.65      0.58        43
weighted avg       0.71      0.63      0.62        43



## The Accuracy we found for the Naive Bayes Algorithm is 63% 